src/patterns.rs

```diff
@@ -20,187 +22,151 @@ use ffi::{
 use ::{
     Path,
     Matrix,
-    MatrixTrait,
     Surface,
 };
 
-// Quite some changes from the C api but all suggested by the cairo devs.
 // See http://cairographics.org/manual/bindings-patterns.html for more info
-#[derive(Debug, Clone)]
-pub enum Pattern {
-    SolidPattern(SolidPattern),
-    SurfacePattern(SurfacePattern),
-    LinearGradient(LinearGradient),
-    RadialGradient(RadialGradient),
-    Mesh(Mesh),
+#[derive(Debug)]
+pub struct Pattern {
+    pointer: *mut cairo_pattern_t,
```

Comment:
`ptr::NonNull`?

Reply:
Should also create an issue to transform any other types to that instead of plain pointers, unless you want to do this as part of this PR too.

Reply:
I think that would work, here and for many (every?) other structs that wrap cairo pointers. But this PR is already large and using `NonNull` internally doesn’t affect the public API, so I’m happy to leave for later or someone else.

Reply:
Works for me :)

Reply:
https://github.com/gtk-rs/cairo/issues/261